src/main/java/de/zalando/aruha/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -167,6 +178,42 @@ public void postEvent(final String topicId, final String partitionId, final Stri
     }
 
     @Override
+    public void syncPostBatch(String topicId, List<BatchItem> batch) throws EventPublishingException {
+        for (BatchItem item : batch) {
+            final ProducerRecord<String, String> record = new ProducerRecord<>(topicId,
+                    toKafkaPartition(item.getPartition()), item.getPartition(), item.getEvent().toString());
+
+            item.setStep(EventPublishingStep.PUBLISHING);
+
+            try {
+                kafkaProducer.send(record, kafkaSendCallback(item));
+            } catch (InterruptException | SerializationException | BufferExhaustedException e) {
+                LOG.error("Error posting event to kafka", e);
+                item.setPublishingStatus(EventPublishingStatus.FAILED);
+                throw new EventPublishingException("Error publishing message to kafka", e);
+            }
+        }
+
+        // this is not the most precise way of waiting for the send to complete, since it will flush all
+        // other concurrent partitions in the accumulator.
+        kafkaProducer.flush();
```

Comment:
This can bring certain problems. As we have a single kafkaProducer instance for the whole application this flush will block not only till all events of this batch are published but till all events that are sent to this kafkaProducer are published. So in a situation when we process 20 different batches at the same time the behaviour here can be quite strange. 

Have you considered a possibility to use another send method that returns future?
`final Future<RecordMetadata> future = kafkaProducer.send(record);`
And then wait till al futures are finished.

(or if you want to stick to callback-implementation - you can then create manual synchronisation with some synchronised counter or something like that)